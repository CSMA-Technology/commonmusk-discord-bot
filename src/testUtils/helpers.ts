/* eslint-disable no-param-reassign */
import {
  Client, CommandInteraction, Guild, TextChannel, ThreadChannel,
} from 'discord.js';

export const createThreadChannel = (
  client: Client,
  guild: Guild,
  interaction: CommandInteraction,
  options?: { threadParentId: string },
) => {
  const threadChannel = Reflect.construct(ThreadChannel, [guild]);
  threadChannel.id = 'ThreadChannel1';
  threadChannel.parentId = options?.threadParentId;
  client.channels.cache.set(threadChannel.id, threadChannel);
  interaction.channelId = threadChannel.id;
  return threadChannel;
};

export const runInThread = (
  client: Client,
  guild: Guild,
  interaction: CommandInteraction,
  run: CommandRunFunc,
  options?: { threadParentId: string },
) => {
  createThreadChannel(client, guild, interaction, options);
  return run(client, interaction);
};

export const runOutsideOfThread = (
  client: Client,
  guild: Guild,
  interaction: CommandInteraction,
  run: CommandRunFunc,
) => {
  const channel = Reflect.construct(TextChannel, [guild, {}, client]);
  channel.id = 'Channel1';
  client.channels.cache.set(channel.id, channel);
  interaction.channelId = channel.id;
  return run(client, interaction);
};

export const convertToMock = (funcs: Function[]) => funcs.map((f) => <jest.Mock>f);
