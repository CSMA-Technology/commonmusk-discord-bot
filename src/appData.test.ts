/* eslint-disable max-len, global-require */
import { existsSync, readFileSync } from 'fs';
import { writeFile } from 'fs/promises';
import { convertToMock } from './testUtils/helpers';

const [mockExistsSync, mockReadFileSync, mockWriteFile] = convertToMock([existsSync, readFileSync, writeFile]);

jest.mock('fs/promises', () => ({
  writeFile: jest.fn(),
}));

jest.mock('fs', () => ({
  existsSync: jest.fn(),
  readFileSync: jest.fn(),
}));

describe('appData', () => {
  const customMetric = { name: 'testMetric', description: 'just a test', type: 'INTEGER' };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should read from a saved file if present', () => {
    jest.isolateModules(() => {
      const serializedAppData = '{"channelMap":{"dataType":"Map","value":[["testChannel","testList"]]},"messageMap":{"dataType":"Map","value":[["testMessage","testCard"]]},"customMetrics":[{"name":"testMetric","description":"just a test","type":"INTEGER"}]}';
      mockExistsSync.mockImplementationOnce(() => true);
      mockReadFileSync.mockImplementationOnce(() => serializedAppData);
      const { channelMap, messageMap, customMetrics } = require('./appData');

      expect(customMetrics.pop()).toEqual(customMetric);

      expect(channelMap.get('testChannel')).toEqual('testList');

      expect(messageMap.get('testMessage')).toEqual('testCard');
    });
  });
  describe('writeAppData', () => {
    it('should serialize and write the config', () => {
      jest.isolateModules(async () => {
        const {
          customMetrics, channelMap, messageMap, writeAppData,
        } = require('./appData');
        customMetrics.push(customMetric);
        channelMap.set('testChannel', 'testList');
        messageMap.set('testMessage', 'testCard');

        await writeAppData();

        expect(mockWriteFile).toHaveBeenCalledTimes(1);
        expect(mockWriteFile.mock.calls[0]).toMatchSnapshot();
      });
    });
  });
});
