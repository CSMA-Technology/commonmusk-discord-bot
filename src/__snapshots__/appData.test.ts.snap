// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appData writeAppData should serialize and write the config 1`] = `
Array [
  "/workspaces/commonmusk-discord-bot/data/appData.json",
  "{\\"channelMap\\":{\\"dataType\\":\\"Map\\",\\"value\\":[[\\"testChannel\\",\\"testList\\"]]},\\"messageMap\\":{\\"dataType\\":\\"Map\\",\\"value\\":[[\\"testMessage\\",\\"testCard\\"]]},\\"customMetrics\\":[{\\"name\\":\\"testMetric\\",\\"description\\":\\"just a test\\",\\"type\\":\\"INTEGER\\"}]}",
]
`;
